File: python/scraper.py
import concurrent.futures
from bs4 import BeautifulSoup
import requests

from utils import create_print_url, generate_html

MAX_WORKERS = 10

class Scraper:
    @staticmethod
    def get_urls_from_list(list_url: str):
        raw_html = requests.get(list_url).text
        soup = BeautifulSoup(raw_html, "html.parser")
        list_element = soup.find(class_="list-links list-musics")
        return [create_print_url(el["href"]) for el in list_element.find_all("a")]

    @staticmethod
    def scrape_page(url: str):
        raw_html = requests.get(url).text
        soup = BeautifulSoup(raw_html, "html.parser")
        content = soup.find(class_="pages")
        return str(content.decode(4, "utf-8"))

    def scrape_pages(self, urls: list[str]):
        with concurrent.futures.ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
            futures = [executor.submit(self.scrape_page, url) for url in urls]
            return [future.result() for future in concurrent.futures.as_completed(futures)]

    def scrape(self, url: str):
        urls = self.get_urls_from_list(url)
        contents = self.scrape_pages(urls)
        return generate_html(contents)

File: typescript/src/scraper.ts
import * as cheerio from 'cheerio';
import { createPrintUrl, generateHtml } from './utils/utils';

export default class Scraper {
    private id: string;
    private total_songs: number;
    private progress: number;

    constructor(_id: string) {
        this.id = _id;
        this.total_songs = 0;
        this.progress = 0;
    }

    private getPage = async (url: string) => {
        const response = await fetch(url);
        return response.text();
    }

    private getCheerioObject = async (url: string) => {
        const html = await this.getPage(url);
        return cheerio.load(html);
    }

    private getUrlsFromList = async (listUrl: string) => {
        const $ = await this.getCheerioObject(listUrl);
        const listObject = $('ol[class="list-links list-musics"]');
        // biome-ignore lint/style/noNonNullAssertion: <explanation>
        return listObject.find('a').map((i, el) => createPrintUrl($(el).attr('href')!)).get();
    }

    private scrapePage = async (url: string) => {
        const $ = await this.getCheerioObject(url);
        const pages = $('div[class="pages"]').get();
        if (pages.length === 0) {
            return '';
        }
        this.progress++;
        const page = pages[0];
        // biome-ignore lint/style/noNonNullAssertion: <explanation>
        return $(page).html()!;
    }

    private scrapePages = async (urls: string[]) => {
        return Promise.all(urls.map(url => this.scrapePage(url)));
    }

    public scrape = async (listUrl: string) => {
        const urls = await this.getUrlsFromList(listUrl);
        this.total_songs = urls.length;
        const htmlContent = await this.scrapePages(urls);
        return generateHtml(htmlContent);
    }
}